src/main/kotlin/com/exactpro/th2/codec/xml/XmlUtils.kt

```
@@ -0,0 +1,83 @@
+        func(item(i))
+    }
+}
+
+fun NodeList.toList(): List<Node> {
+    val list = ArrayList<Node>(this.length)
+
+    this.forEach {
+        list.add(it)
+    }
+    return list
+}
+
+fun Node.getText(): String =
+    nodeValue ?: childNodes.let {
+        if (it.length == 0) "" else it.toList().find { node -> node.nodeType == Node.TEXT_NODE }?.nodeValue ?: ""
+    }
+
+fun Node.setText(text: String, document: Document) {
+    appendChild(document.createTextNode(text))
+}
+
+fun Node.addNode(name: String, document: Document) : Node = appendChild(document.createElement(name))
+
+fun IDictionaryStructure.getCharsetName(): String? = attributes[XmlPipelineCodec.XML_CHARSET_ATTRIBUTE]?.value
+
+fun IFieldStructure.getXmlName(): String = attributes[XmlPipelineCodec.XML_TAG_NAME_ATTRIBUTE]?.value ?: name
+fun IFieldStructure.getXPathExpression(): String? = attributes[XmlPipelineCodec.XML_X_PATH_EXPRESSION_ATTRIBUTE]?.value
+fun IFieldStructure.getAttrName(): String? = attributes[XmlPipelineCodec.XML_ATTRIBUTE_NAME_ATTRIBUTE]?.value
+fun IFieldStructure.isValidNode(node: Node): Boolean {
+    if (node.nodeType != Node.ELEMENT_NODE) {
+        return false
+    }
+
+    for (i in 0 until node.attributes.length) {
+        val attr = node.attributes.item(i)
+        if (this.attributes[XmlPipelineCodec.XML_ATTRIBUTE_NAME_ATTRIBUTE]?.value == attr.nodeName
+            && this.attributes[XmlPipelineCodec.XML_ATTRIBUTE_VALUE_ATTRIBUTE]?.value == attr.nodeValue
+        ) {
+            return true
+        }
+    }
+
+    return node.nodeName == this.getXmlName()
+}
+
+fun IMessageStructure.isEmbedded() : Boolean = attributes[XmlPipelineCodec.EMBEDDED_ATTRIBUTE]?.getCastValue() ?: false
+
+fun XPath.find(expression: String, parentNode: Node, handleError: (Exception) -> Exception): NodeList {
+    try {
+        return this.evaluate(expression, parentNode, XPathConstants.NODESET) as NodeList
+    } catch (e: Exception) {
+        throw handleError(e)
+    }
+}
```

Should we throw an exception when a field isn't required?

---

If nodes did not find, this method returns NodeList without nodes. We catch exception if get the wrong XPath expression or set the wrong result type.